### PR TITLE
[Contrib] Checkout a PR

### PR DESCRIPTION
**Pull request by sapk**
Could be impoved, but it seems to work.

For example, to check the PR #6007 simply do :
```go run contrib/pr/checkout.go 6007```
OR
```make pr PR=6007```

This will checkout PR and start with integration tests fixtures loaded.

There still seems to have problem with auth after login the user is not kept logged (and some repo are missing but that normal)

### CODE REVIEW

**Review by lafriks — APPROVED** [review]

**Comment by zeripath** [comment]
@sapk do you think you can fix the issues you mentioned?

**Comment by sapk** [comment]
@zeripath Since it is mostly to ease validating UI PR and have some data to view, I didn't take time on that part. I think it is related to fixtures but can't find exactly why.

**Comment by lafriks** [comment]
@zeripath I think fixing fixtures should be out of scope for this PR

**Comment by zeripath** [comment]
I was meaning about keeping the user logged in, but if it's too hard to fix then we can ignore it as it's only contrib

**Review by jolheiser — COMMENTED** [review]
~~I couldn't for the life of me figure out how to get it to retain auth, but~~ the rest of these nits just came about as I used the contrib. I don't think any of my nits are blocking though, so I'll just leave them as comments.
It seems to work well enough otherwise. 😄 
I will say this makes checking out PRs incredibly easy!

File: contrib/pr/checkout.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	models.LoadConfigs()
+	routers.NewServices()
+	//x, err = xorm.NewEngine("sqlite3", "file::memory:?cache=shared")
+
+	var helper testfixtures.Helper
+	helper = &testfixtures.SQLite{}
+	models.NewEngine(func(_ *xorm.Engine) error {
+		return nil
+	})
+	//x.ShowSQL(true)
```
Comment (jolheiser) [inline contrib/pr/checkout.go:97]: To fix auth, after initializing the engine here, add `models.HasEngine = true`

File: contrib/pr/checkout.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+	//Copy this file if it will not exist in the PR branch
+	dat, err := ioutil.ReadFile(codeFilePath)
+	if err != nil {
+		log.Fatalf("Failed to cache this code file : %v", err)
+	}
+
+	branch := fmt.Sprintf("pr-%s-%d", pr, time.Now().Unix())
+	log.Printf("Checkout PR #%s in %s\n", pr, branch)
+	runCmd("git", "fetch", "origin", fmt.Sprintf("pull/%s/head:%s", pr, branch))
```
[the remark this answers is not in the document]
Comment (sapk) [reply contrib/pr/checkout.go]: I will try to list remotes and find the one wich is gitea upstream. ;-)

File: contrib/pr/checkout.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+		}
+
+		if setting.EnablePprof {
+			go func() {
+				log.Info("Starting pprof server on localhost:6060")
+				log.Info("%v", http.ListenAndServe("localhost:6060", nil))
+			}()
+		}
+	*/
+	gracehttp.Serve(&http.Server{
```
[the remark this answers is not in the document]
Comment (sapk) [reply contrib/pr/checkout.go]: I will migrate to native  http.ListenAndServe as grace is not really needed for testing/validating PR.

File: contrib/pr/checkout.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+	com.CopyDir(path.Join(curDir, "integrations/gitea-repositories-meta"), setting.RepoRootPath)
+
+	log.Printf("[PR] Setting up router\n")
+	//routers.GlobalInit()
+	external.RegisterParsers()
+	m := routes.NewMacaron()
+	routes.RegisterRoutes(m)
+
+	log.Printf("[PR] Ready for testing !\n")
+	log.Printf("[PR] Login with user1, user2, user3, ... with pass: passsword\n")
```
Comment (jolheiser) [inline contrib/pr/checkout.go]: Last thing from me I think. The world's tiniest nit. `password` instead of `passsword`?

**Comment by sapk** [comment]
Thanks to @jolheiser the auth is working (ex: user1:password). This will also try to find the remote corresponding to gitea upstream and not force to origin (still the default in case of failure). And should also work on windows (not tested).

File: contrib/pr/checkout.go (hunk 4 above)
Comment (sapk) [reply contrib/pr/checkout.go]: Thanks for the catch.

File: contrib/pr/checkout.go (hunk 1 above)
Comment (sapk) [reply contrib/pr/checkout.go:97]: Thanks for the solution.

File: contrib/pr/checkout.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sapk) [reply contrib/pr/checkout.go]: Done.

File: contrib/pr/checkout.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	}
+	if err := c.Start(); err != nil {
+		log.Panicln(err)
+	}
+	if err := c.Wait(); err != nil {
+		log.Panicln(err)
+	}
+	var b []byte
+	stdout.Read(b)
+	log.Println("Log: " + string(b)) //TODO stream to output
```
[the remark this answers is not in the document]
Comment (sapk) [reply contrib/pr/checkout.go]: Thanks

**Comment by jolheiser** [comment]
When I tested on Windows the only hangup was grace, so this should work splendidly. I plan on using this for reviewing a ton, thanks for adding it!

File: contrib/pr/checkout.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sapk) [reply contrib/pr/checkout.go]: Done

**Review by jolheiser — APPROVED** [review]

**Comment by sapk** [comment]
@jolheiser You are doing great code review. The more easy to do it the more gitea code-review will be thorough.

**Comment by jolheiser** [comment]
After testing this again with latest changes, now it seems to be complaining about unstaged changes, even though I have a clean working tree. Did you run into this at all?

EDIT: After changing the checkout method to 
```go
err = tree.Checkout(&git.CheckoutOptions{
	Branch: plumbing.ReferenceName(branch),
	Force: true,
})
```
it worked...not sure if an issue with my setup or if it is needed for some reason. It *shouldn't* need to force the reset, but perhaps something else in the code is causing a "dirty" working tree before checkout?

**Comment by lafriks** [comment]
Would be nice to add shortcut to makefile for this, something like `make pr 1234`

**Comment by sapk** [comment]
@jolheiser I did run it but it seems to not really create the branch at checkout but only checkouting the ref.
The unstaged changes can maybe come from the fact that the source file is duplicated in the destination repo if it doesn't exist in this revision. I will look into it.

**Comment by sapk** [comment]
@jolheiser did you use the command on windows ? 
On windows : I have the error message of unstaged changes
On linux : It doesn't complain and checkout the PR and to the rest but it doesn't create the branch that it should at fetch step.

I will debug more to find why.

**Comment by sapk** [comment]
I find the mistake for the linux but still have the error on windows. The current solution is to use `Force: runtime.GOOS == "windows"` to only force on windows. (like suggested by @jolheiser)

@lafriks To have a make target like that we should had some hacky code like :
```
ifeq (pr, $(firstword $(MAKECMDGOALS)))
  runargs := $(wordlist 2, $(words $(MAKECMDGOALS)), $(MAKECMDGOALS))
  $(eval $(runargs):;@true)
endif
.PHONY: pr
pr:
	$(GO) run contrib/pr/checkout.go $(runargs)
```
Or we could also use format like `make pr PR=0000`
```
.PHONY: pr
pr:
    (GO) run contrib/pr/checkout.go $(PR)
```

**Comment by jolheiser** [comment]
I had been testing on Windows at the time, yes. Good to know it's resolved on Linux, as that's where I will probably be doing most of my dev work going forward.   

The only issue I can think of is if `Force` would discard working changes the dev wanted but forgot to commit before using the contrib. (Although we have fairly few Windows devs afaik)  
  
As for the Makefile, personally I think the second option looks cleaner, and I like the idea of not making the Makefile any messier.

**Comment by lafriks** [comment]
I'm ok with second option for makefile

File: Makefile [hunk 6]
```diff
@@ -420,3 +420,7 @@ generate-images:
 					$(TMPDIR)/images/64.png $(TMPDIR)/images/128.png \
 					$(PWD)/public/img/favicon.ico
 	rm -rf $(TMPDIR)/images
+	
+.PHONY: pr
+pr:
+    $(GO) run contrib/pr/checkout.go $(PR)
```
Comment (jolheiser) [inline Makefile]: Regarding failed CI, I think Makefile is just complaining about this not being a tab.
```suggestion
	$(GO) run contrib/pr/checkout.go $(PR)
```

Comment (sapk) [reply Makefile]: Yes, I edit it via github web interface and it fail ... I was waiting to have time to fix this but thanks for the correction.

Comment (jolheiser) [reply Makefile]: Hah sorry, I'm just really excited to be using this. 😄

**Pull request merged**